src/com/hrabrov/introduction_to_java/algorithmization/multidimensional_array/Task16.java

```diff
@@ -0,0 +1,229 @@
+package com.hrabrov.introduction_to_java.algorithmization.multidimensional_array;
+
+public class Task16 {
+    /**
+     * Построить магический квадрат для матрицы размера n.
+     * Generate magic square for matrix of size n.
+     *
+     * @param n input size of magic square
+     * @return multidimensional array like magic square
+     */
+
+    public static int[][] generateMagicSquare(int n) {
+        if (n % 2 != 0) {
+            return generateMagicSquareForOddSize(n);
+        } else if (n % 4 == 0) {
+            return generateMagicSquareForEvenEvenSize(n);
+        }
+
+        return generateMagicSquareForEvenOddSize(n);
+    }
+
+    //n = 2k + 1, k = 1, 2, 3, ...
+    private static int[][] generateMagicSquareForOddSize(int n) {
+        int middleOfSideOfMagicSquare;
+        middleOfSideOfMagicSquare = findMiddleOfSideOfMatrix(n);
+
+        //create auxiliary matrix (bigMatrix)
+        int howManyCellToAdd;
+        howManyCellToAdd = (n - middleOfSideOfMagicSquare) * 2;
+        int[][] auxiliaryBigMatrix;
+        int sizeOfAuxiliaryBigMatrix;
+        sizeOfAuxiliaryBigMatrix = n + howManyCellToAdd;
+        auxiliaryBigMatrix = new int[sizeOfAuxiliaryBigMatrix][sizeOfAuxiliaryBigMatrix];
+
+        for (int i = 0; i < sizeOfAuxiliaryBigMatrix; i++) {
+            for (int j = 0; j < sizeOfAuxiliaryBigMatrix; j++) {
+                auxiliaryBigMatrix[i][j] = 0;
+            }
+        }
+
+        //fill in diagonal rows
+        int middleOfSideOfAuxiliaryBigMatrix;
+        middleOfSideOfAuxiliaryBigMatrix = findMiddleOfSideOfMatrix(sizeOfAuxiliaryBigMatrix);
+        int numberForFillAuxiliaryBigMatrix = 1;
+        int tempJ = 0;
+        for (int i = middleOfSideOfAuxiliaryBigMatrix - 1; i < sizeOfAuxiliaryBigMatrix; i++) {
+            int tempI = i;
+            for (int j = tempJ; j <= i; j++) {
+                auxiliaryBigMatrix[tempI][j] = numberForFillAuxiliaryBigMatrix;
+                numberForFillAuxiliaryBigMatrix++;
+                tempI -= 1;
+            }
+            tempJ++;
+        }
+
+        //transfer number from the terrace to the magic square
+        int halfOfAddedCells;
+        halfOfAddedCells = howManyCellToAdd / 2;
+        for (int i = 0; i < sizeOfAuxiliaryBigMatrix; i++) {
+            for (int j = 0; j < sizeOfAuxiliaryBigMatrix; j++) {
+                //top terrace
+                if (i < halfOfAddedCells) {
+                    if (auxiliaryBigMatrix[i][j] != 0) {
+                        auxiliaryBigMatrix[i + n][j] = auxiliaryBigMatrix[i][j];
+                        auxiliaryBigMatrix[i][j] = 0;
+                    }
```

Comment:
I think this can be extracted to separate method to avoid copy-pasting of almost the same code over and over